RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+      }
+  }
+
+  // Seeding using local energy maxima
+  class SeedingTopoThresh {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  const reco::PFClusterParamsDeviceCollection::ConstView pfClusParams,
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  reco::PFClusterDeviceCollection::View clusterView,
+                                  reco::PFRecHitFractionDeviceCollection::View fracView,
+                                  uint32_t* __restrict__ nSeeds) const {
+      const int nRH = pfRecHits.size();
+      int i = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] +
+              alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u] *
+                  alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];
+      clusterView.size() = nRH;
+
+      if (i < nRH) {
+        // Initialize arrays
+        pfClusteringVars[i].pfrh_topoId() = i;
+        pfClusteringVars[i].pfrh_isSeed() = 0;
+        pfClusteringVars[i].rhCount() = 0;
+        pfClusteringVars[i].topoSeedCount() = 0;
+        pfClusteringVars[i].topoRHCount() = 0;
+        clusterView[i].topoRHCount() = 0;
+        pfClusteringVars[i].seedFracOffsets() = -1;
+        pfClusteringVars[i].topoSeedOffsets() = -1;
+        pfClusteringVars[i].topoSeedList() = -1;
+        clusterView[i].seedRHIdx() = -1;
+
+        int layer = pfRecHits[i].layer();
+        int depthOffset = pfRecHits[i].depth() - 1;
+        float energy = pfRecHits[i].energy();
+        Position3 pos = Position3{pfRecHits[i].x(), pfRecHits[i].y(), pfRecHits[i].z()};
+
+        // cmssdt.cern.ch/lxr/source/DataFormats/ParticleFlowReco/interface/PFRecHit.h#0108
+        float pt2 = energy * energy * (pos.x * pos.x + pos.y * pos.y) / (pos.x * pos.x + pos.y * pos.y + pos.z * pos.z);
+
+        // Seeding threshold test
+        if ((layer == PFLayer::HCAL_BARREL1 && energy > pfClusParams.seedEThresholdEB_vec()[depthOffset] &&
+             pt2 > pfClusParams.seedPt2ThresholdEB()) ||
+            (layer == PFLayer::HCAL_ENDCAP && energy > pfClusParams.seedEThresholdEE_vec()[depthOffset] &&
+             pt2 > pfClusParams.seedPt2ThresholdEE())) {
+          pfClusteringVars[i].pfrh_isSeed() = 1;
+          for (int k = 0; k < 4; k++) {  // Does this seed candidate have a higher energy than four neighbours
+            if (pfRecHits[i].neighbours()(k) < 0)
+              continue;
+            if (energy < pfRecHits[pfRecHits[i].neighbours()(k)].energy()) {
+              pfClusteringVars[i].pfrh_isSeed() = 0;
+              break;
+            }
+          }
+          if (pfClusteringVars[i].pfrh_isSeed())
+            alpaka::atomicAdd(acc, nSeeds, 1u);
```

Comment:
```suggestion
            alpaka::atomicInc(acc, nSeeds, std::numeric_limits<uint32_t>::max());
```

`atomicInc` may have better performance than `atomicAdd`.

Reply:
Thank you. At least to me, the current one is easier to understand :).